mcnntunes errors for PerBin Model

michaeldessena opened the pull request:
Hi Marco,
I hope I have done it in the right way. Let me know if there is something wrong. I think we have to delete all my useless comments from the code to get it clear.

michaeldessena committed:
add mcnntunes_buildruns2.py

michaeldessena committed:
Add errors evaluation

michaeldessena committed:
install with pip

michaeldessena committed:
install with pip

michaeldessena committed:
chi2 + Qgamma

michaeldessena committed:
add to report

michaeldessena committed:
chi2/dof=chi2/dof_min+Q/dof

michaeldessena committed:
dof chi2=nbins-n params

michaeldessena committed:
Inverse model riparazione

michaeldessena committed:
check

michaeldessena committed:
correction

mlazzarin commented:
Thanks, I'm going to review the changes. For now, please remove all files in the ``mcnntunes.egg-info`` folder.

mlazzarin commented:
I believe you can also remove the commented lines of code, so that we minimize the changes. If needed, we can always recover them by going back to the previous commits.

michaeldessena commented:
Ok, thanks. Only one question after that I have removed the file and the
comments I have to make another pull request (once I have updated my
brench)?

Il ven 22 apr 2022, 10:31 Marco Lazzarin ***@***.***> ha
scritto:

> I believe you can also remove the commented lines of code, so that we
> minimize the changes. If needed, we can always recover them by going back
> to the previous commits.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/N3PDF/mcnntunes/pull/24#issuecomment-1106188384>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AOUMDMJOJHLMLDS3J6QZDCTVGJPUHANCNFSM5T6T77RQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

mlazzarin commented:
No, you can just push the commits to your fork (like you did before) and this PR will be automatically updated.

michaeldessena committed:
comments removed

michaeldessena commented:
Ok, done! Let me know if is it okay

mlazzarin committed:
Fix typo

mlazzarin committed:
Fix index.html for Inverse Model

mlazzarin commented:
> Ok, done! Let me know if is it okay

Thanks! I will review everything and propose some changes

mlazzarin committed:
Minor fixes

mlazzarin committed:
Beginning refactoring of PerBin model errors

mlazzarin committed:
Further refactoring

mlazzarin committed:
Complete refactor of PerBinModel errors

mlazzarin committed:
Fix previous commit

mlazzarin committed:
Further fixes

mlazzarin committed:
Fixes

mlazzarin committed:
Refactor dof for chi2

michaeldessena committed:
Merge pull request #1 from mlazzarin/minimize

Refactor error computation for PerBin model

mlazzarin commented:
Also, given that this PR involves the computation of the errors only, I believe it's better to remove the additional script (buildruns2), and maybe leave it for future contributions. Do you agree?

mlazzarin committed:
Make dof consistent everywhere

mlazzarin committed:
Bugfix

michaeldessena committed:
Merge pull request #2 from mlazzarin/fixdof

Make dof consistent everywhere

michaeldessena committed:
remove mcnntunes_buildruns2.py

michaeldessena commented:
Ok I removed the mcnntunes_buildruns2.py script

mlazzarin committed:
Add PR trigger for CI

mlazzarin committed:
Merge remote-tracking branch 'n3pdf/master'

mlazzarin committed:
Fix bug in benchmark mode with asymmetric errors

mlazzarin committed:
Merge remote-tracking branch 'n3pdf/master'

mlazzarin commented:
Ok, all checks have passed. @michaeldessena I believe we are ready for review, do you agree?

mlazzarin committed:
Add dof to chi2 Keras loss

mlazzarin commented on src/mcnntunes/minimizer.py:
```diff
@@ -31,22 +32,65 @@ def __init__(self, runs, truth, model, output = None, truth_index = 0):
+                chi2_values[index] = self.chi2(a)
+                x_unscaled[index]  = self.runs.unscale_x(a)[axis]
+
+            # Compute where chi2 - min(chi2) ~= delta_chi2
+            intersec_idxs = np.argwhere(np.diff(np.sign(
+                                chi2_values - np.min(chi2_values) - delta_chi2
+                            )))
```
Possible caveat: if the ``min(chi2_values)`` is not equal to the minimum found by CMAES, we would obtain a weird result. However, this would underline a failure in the minimization process.

mlazzarin commented:
> Ok, what should I do? sorry but i'm not an expert on github I have used it only for school projects.

I'm going to ask @scarrazza for review, then we can merge.


Changes:
 - Made degrees of freedom more consistent in the code:
    - Number of bins - number of tuning params (unweighted case)
    - Sum of weights - number of tuning params (weighted case)
 - Implemented errors for PerBin model by searching for the interval in which ``chi2 - min(chi2)`` falls in the 69 percentile of chi2 distribution. 

Everything looks good to me, I found a possible caveat (I left a comment) but I believe it's negligible.
I don't have a theoretical basis for the computation of the weighted dof, but it kind of works (well, if you set the weights to 1000, the number of params becomes negligible)

scarrazza reviewed (commented):
Looks good to me. Did you or Michael try to run a benchmark fit to confirm everything is working as expected?

michaeldessena commented:
I've tried it on my master thesis analysis, seem work properly

scarrazza commented:
Ok, thanks, we can merge them.
